Database-Features/view-optimization.md

```diff
@@ -0,0 +1,165 @@
+
+### Examples
+
+Notes:
+
+- all query rewrites are performed on the compiled query graph, not in SQL text as shown below.
+- all queries are based on a TPC-DS data set of 10 GB (scale factor 10). The TPD-DS data generator always creates data for the years 1998 through 2003.
+
+```sql
+-- original query: materialized view would contain 7 million rows
+WITH big_sales AS (
+    SELECT * FROM catalog_sales
+    WHERE cs_quantity > 50
+)
+SELECT 
+    d_year, SUM(cs_net_paid) AS turnover
+FROM big_sales
+JOIN date_dim
+  ON d_date_sk = cs_sold_date_sk
+GROUP BY d_year;
+
+-- after subselect elimination: no intermediate materialization
+SELECT 
+    d_year, SUM(cs_net_paid) AS turnover
+FROM catalog_sales
+JOIN date_dim
+  ON d_date_sk = cs_sold_date_sk
+WHERE cs_quantity > 50
+GROUP BY d_year;
+```
+
+```sql
+-- original query: materialized view would process 14 million records
+WITH yearly_turnover AS (
+    SELECT 
+        d_year, SUM(cs_net_paid) AS turnover
+    FROM catalog_sales
+    JOIN date_dim
+      ON d_date_sk = cs_sold_date_sk
+   GROUP BY d_year
+)
+SELECT SUM(turnover)
+FROM yearly_turnover
+WHERE d_year >= 2002;
+
+-- after filter propagation: processes 3 million records
+WITH yearly_turnover AS (
+    SELECT 
+        d_year, SUM(cs_net_paid) AS turnover
+    FROM catalog_sales
+    JOIN date_dim
+      ON d_date_sk = cs_sold_date_sk
+   WHERE d_year >= 2002
+   GROUP BY d_year
+)
+SELECT SUM(turnover)
+FROM yearly_turnover
+;
+```
+
+## Multiple Usages: View Optimization
+
+The mechanics above are usually pretty straight-forward and rarely cause problems.
+
+However, things get more complicated when a view is **used multiple times** within a statement:
+
+### Example
+
+```sql
+WITH yearly_turnover AS (
+    SELECT 
+        d_year, SUM(cs_net_paid) AS turnover
+    FROM catalog_sales
+    JOIN date_dim
+      ON d_date_sk = cs_sold_date_sk
+   GROUP BY d_year
+)
+SELECT 'previous' AS "YEAR", SUM(turnover) AS turnover
+  FROM yearly_turnover WHERE d_year = 2002
+UNION ALL
+SELECT 'current', SUM(turnover)
+  FROM yearly_turnover WHERE d_year = 2003
+;
+```
+
+The optimizer now has to decide between the following two options. This decision is called **view optimization** at Exasol.
+
+### 1 - Materialize the view once
+
+...and reuse its data in all occurrences. This will guarantee identical data in all places, but might cause excessive RAM usage if the view returns too much &mdash; potentially unused &mdash; data.
+
+In the example above, this would again cause processing of 14 million rows, followed by two very quick filters.
+
+### 2 - Treat each occurrence separately
+
+...and apply the mechanisms outlined in [Background](#background-other-optimizations) above. While this will usually result in much smaller materializations (or none at all), it will potentially incur duplicate processing efforts and might &mdash; at least in theory &mdash; return inconsistent data due to unpredictable rounding errors in each evaluation.
```

Comment:
Let's see if this links is going to be rendered.
Say, I don't see it tested in "Extended Markdown Formatting" article.

Reply:
@allipatev it does get rendered, but it won't work within salesforce.